### ISSUE TITLE
Passing a scalar torch.Tensor to an IntList is completely broken

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#167562]
### 🐛 Describe the bug

PythonArgParser accepts a scalar Tensor in places that expect an `IntList`. However, the parsing code casts the Tensor to a PyListObject. It just so happens that PyListObject `ob_size` is the same place as THPVariable's `bool isBorrowed_` from `MaybeOwned<Tensor>` so the parsing code thinks it's an empty list (!?) but none of this makes any sense.

For example, in eager mode:

```python
    x = torch.ones(1, 2, 3, 4, 5)
    print(x.sum(dim=torch.tensor(3)))  # Ed, why are we accepting a scalar tensor here?!
    print(x.sum(dim=3))
```

It also shows up in places where we generate scalar FakeTensors, such as `random.randint()`:

```python
import torch
import random

def foo():
    x = torch.ones(2, 2, 2)
    return x.sum(dim=random.randint(0, 0))  # i.e.,  dim=0

@torch.compile
def foo_compile():
    x = torch.ones(2, 2, 2)
    return x.sum(dim=random.randint(0, 0))  # i.e.,  dim=0

assert foo().shape == foo_compile().shape
```

cc @albanD @ezyang

### Versions

It's on main as of Nov 11, 2025 (fad48ffa62e1e0a39ee52b649b7c27949772a965)

**Comment by ezyang** [comment]
Checking if Claude can find the exact line that needs to be re-reviewed.

**Comment by ezyang** [comment]
Claude gave botslop. The root cause is that sum's dim intlist is broadcasted `IntList[1]` which puts us into the int path, which passes it due to this:

```
  // FakeTensor(..., size=()) is qualified for SymInt param,
  // but we can't go via __index__ (below) as we would normally
  // do for regular tensors, because __index__ first forces a
  // conversion into an int, which in general you cannot do
  // if you have an unbacked SymInt.  So this fastpath ensures
  // that we still allow for fake tensors in this case, but
  // for regular tensors it's redundant with the test below.
  if (THPVariable_Check(obj)) {
    auto& var = THPVariable_Unpack(obj);
    if (TORCH_GUARD_OR_FALSE(var.sym_numel().sym_eq(1)) &&
        at::isIntegralType(var.dtype().toScalarType(), /*include_bool*/ true)) {
      return true;
    }
  }
```

**Comment by ezyang** [comment]
I think we're supposed to accept the tensor here anyway, because tensor has `__index__` implemented and we try to do this in the last ditch resort. So I think we just need to update the list parsing path to handle this too.
